:AllowTypeErrors

def succ: int -> int
//│ succ: int -> int


x1 = false
x2 = x1
x3 = x2
//│ x1: false
//│ x2: false
//│ x3: false

:ex
succ x3
//│ ╔══[ERROR] Type mismatch in application:
//│ ║  l.15: 	succ x3
//│ ║        	^^^^^^^
//│ ╟── reference of type `false` is not an instance of type `int`
//│ ║  l.7: 	x1 = false
//│ ║       	     ^^^^^
//│ ╟── but it flows into reference with expected type `int`
//│ ║  l.15: 	succ x3
//│ ║        	     ^^
//│ ╟── Note: constraint arises from type reference:
//│ ║  l.3: 	def succ: int -> int
//│ ║       	          ^^^
//│ ╟── ========= Additional explanations below =========
//│ ╟── [info] flowing from type `false`
//│ ╟── [info] flowing from reference of type `false`
//│ ║  l.7: 	x1 = false
//│ ║       	     ^^^^^
//│ ╟── [info] flowing from reference of type `false`
//│ ║  l.8: 	x2 = x1
//│ ║       	     ^^
//│ ╟── [info] flowing from reference of type `false`
//│ ║  l.9: 	x3 = x2
//│ ║       	     ^^
//│ ╟── [info] flowing from reference of type `false`
//│ ║  l.15: 	succ x3
//│ ║        	     ^^
//│ ╟── [info] flowing into type `int`
//│ ║  l.3: 	def succ: int -> int
//│ ║       	          ^^^
//│ ╟── [info] flowing into type `int`
//│ ║  l.3: 	def succ: int -> int
//│ ╙──     	          ^^^
//│ res: error | int


f1 y1 = succ y1
f2 y2 = f1 y2
f3 y3 = f2 y3
//│ f1: int -> int
//│ f2: int -> int
//│ f3: int -> int

:ex
f3 true
//│ ╔══[ERROR] Type mismatch in application:
//│ ║  l.59: 	f3 true
//│ ║        	^^^^^^^
//│ ╟── reference of type `true` is not an instance of type `int`
//│ ║  l.59: 	f3 true
//│ ║        	   ^^^^
//│ ╟── Note: constraint arises from type reference:
//│ ║  l.3: 	def succ: int -> int
//│ ║       	          ^^^
//│ ╟── from reference:
//│ ║  l.53: 	f3 y3 = f2 y3
//│ ║        	           ^^
//│ ╟── ========= Additional explanations below =========
//│ ╟── [info] flowing from type `true`
//│ ╟── [info] flowing from reference of type `true`
//│ ║  l.59: 	f3 true
//│ ║        	   ^^^^
//│ ╟── [info] flowing into variable of type `int`
//│ ╟── [info] flowing into reference of type `int`
//│ ║  l.53: 	f3 y3 = f2 y3
//│ ║        	           ^^
//│ ╟── [info] flowing into variable of type `int`
//│ ╟── [info] flowing into reference of type `int`
//│ ║  l.52: 	f2 y2 = f1 y2
//│ ║        	           ^^
//│ ╟── [info] flowing into variable of type `int`
//│ ╟── [info] flowing into reference of type `int`
//│ ║  l.51: 	f1 y1 = succ y1
//│ ║        	             ^^
//│ ╟── [info] flowing into type `int`
//│ ║  l.3: 	def succ: int -> int
//│ ║       	          ^^^
//│ ╟── [info] flowing into type `int`
//│ ║  l.3: 	def succ: int -> int
//│ ║       	          ^^^
//│ ╟── [info] flowing into type `int`
//│ ║  l.3: 	def succ: int -> int
//│ ║       	          ^^^
//│ ╟── [info] flowing into type `int`
//│ ║  l.3: 	def succ: int -> int
//│ ║       	          ^^^
//│ ╟── [info] flowing into type `int`
//│ ║  l.3: 	def succ: int -> int
//│ ╙──     	          ^^^
//│ res: error | int


:ex
f3 x3
//│ ╔══[ERROR] Type mismatch in application:
//│ ║  l.108: 	f3 x3
//│ ║         	^^^^^
//│ ╟── reference of type `false` is not an instance of type `int`
//│ ║  l.7: 	x1 = false
//│ ║       	     ^^^^^
//│ ╟── but it flows into reference with expected type `int`
//│ ║  l.108: 	f3 x3
//│ ║         	   ^^
//│ ╟── Note: constraint arises from type reference:
//│ ║  l.3: 	def succ: int -> int
//│ ║       	          ^^^
//│ ╟── from reference:
//│ ║  l.53: 	f3 y3 = f2 y3
//│ ║        	           ^^
//│ ╟── ========= Additional explanations below =========
//│ ╟── [info] flowing from type `false`
//│ ╟── [info] flowing from reference of type `false`
//│ ║  l.7: 	x1 = false
//│ ║       	     ^^^^^
//│ ╟── [info] flowing from reference of type `false`
//│ ║  l.8: 	x2 = x1
//│ ║       	     ^^
//│ ╟── [info] flowing from reference of type `false`
//│ ║  l.9: 	x3 = x2
//│ ║       	     ^^
//│ ╟── [info] flowing from reference of type `false`
//│ ║  l.108: 	f3 x3
//│ ║         	   ^^
//│ ╟── [info] flowing into variable of type `int`
//│ ╟── [info] flowing into reference of type `int`
//│ ║  l.53: 	f3 y3 = f2 y3
//│ ║        	           ^^
//│ ╟── [info] flowing into variable of type `int`
//│ ╟── [info] flowing into reference of type `int`
//│ ║  l.52: 	f2 y2 = f1 y2
//│ ║        	           ^^
//│ ╟── [info] flowing into variable of type `int`
//│ ╟── [info] flowing into reference of type `int`
//│ ║  l.51: 	f1 y1 = succ y1
//│ ║        	             ^^
//│ ╟── [info] flowing into type `int`
//│ ║  l.3: 	def succ: int -> int
//│ ║       	          ^^^
//│ ╟── [info] flowing into type `int`
//│ ║  l.3: 	def succ: int -> int
//│ ║       	          ^^^
//│ ╟── [info] flowing into type `int`
//│ ║  l.3: 	def succ: int -> int
//│ ║       	          ^^^
//│ ╟── [info] flowing into type `int`
//│ ║  l.3: 	def succ: int -> int
//│ ║       	          ^^^
//│ ╟── [info] flowing into type `int`
//│ ║  l.3: 	def succ: int -> int
//│ ╙──     	          ^^^
//│ res: error | int


h1 f x = f x
h2 f x = h1 f x
h3 f x = h2 f x
//│ h1: ('a -> 'b) -> 'a -> 'b
//│ h2: ('a -> 'b) -> 'a -> 'b
//│ h3: ('a -> 'b) -> 'a -> 'b

:ex
h3 f3 x3
//│ ╔══[ERROR] Type mismatch in application:
//│ ║  l.176: 	h3 f3 x3
//│ ║         	^^^^^^^^
//│ ╟── reference of type `false` is not an instance of type `int`
//│ ║  l.7: 	x1 = false
//│ ║       	     ^^^^^
//│ ╟── but it flows into reference with expected type `int`
//│ ║  l.176: 	h3 f3 x3
//│ ║         	      ^^
//│ ╟── Note: constraint arises from type reference:
//│ ║  l.3: 	def succ: int -> int
//│ ║       	          ^^^
//│ ╟── from reference:
//│ ║  l.170: 	h3 f x = h2 f x
//│ ║         	              ^
//│ ╟── ========= Additional explanations below =========
//│ ╟── [info] flowing from type `false`
//│ ╟── [info] flowing from reference of type `false`
//│ ║  l.7: 	x1 = false
//│ ║       	     ^^^^^
//│ ╟── [info] flowing from reference of type `false`
//│ ║  l.8: 	x2 = x1
//│ ║       	     ^^
//│ ╟── [info] flowing from reference of type `false`
//│ ║  l.9: 	x3 = x2
//│ ║       	     ^^
//│ ╟── [info] flowing from reference of type `false`
//│ ║  l.176: 	h3 f3 x3
//│ ║         	      ^^
//│ ╟── [info] flowing into variable of type `?a`
//│ ╟── [info] flowing into reference of type `?a`
//│ ║  l.170: 	h3 f x = h2 f x
//│ ║         	              ^
//│ ╟── [info] flowing into variable of type `?a`
//│ ╟── [info] flowing into reference of type `?a`
//│ ║  l.169: 	h2 f x = h1 f x
//│ ║         	              ^
//│ ╟── [info] flowing into variable of type `?a`
//│ ╟── [info] flowing into variable of type `?a`
//│ ╟── [info] flowing into variable of type `?a`
//│ ╟── [info] flowing into variable of type `int`
//│ ╟── [info] flowing into reference of type `int`
//│ ║  l.168: 	h1 f x = f x
//│ ║         	           ^
//│ ╟── [info] flowing into variable of type `int`
//│ ╟── [info] flowing into reference of type `int`
//│ ║  l.53: 	f3 y3 = f2 y3
//│ ║        	           ^^
//│ ╟── [info] flowing into variable of type `int`
//│ ╟── [info] flowing into reference of type `int`
//│ ║  l.52: 	f2 y2 = f1 y2
//│ ║        	           ^^
//│ ╟── [info] flowing into variable of type `int`
//│ ╟── [info] flowing into reference of type `int`
//│ ║  l.51: 	f1 y1 = succ y1
//│ ║        	             ^^
//│ ╟── [info] flowing into type `int`
//│ ║  l.3: 	def succ: int -> int
//│ ║       	          ^^^
//│ ╟── [info] flowing into type `int`
//│ ║  l.3: 	def succ: int -> int
//│ ║       	          ^^^
//│ ╟── [info] flowing into type `int`
//│ ║  l.3: 	def succ: int -> int
//│ ║       	          ^^^
//│ ╟── [info] flowing into type `int`
//│ ║  l.3: 	def succ: int -> int
//│ ║       	          ^^^
//│ ╟── [info] flowing into type `int`
//│ ║  l.3: 	def succ: int -> int
//│ ║       	          ^^^
//│ ╟── [info] flowing into type `int`
//│ ║  l.3: 	def succ: int -> int
//│ ╙──     	          ^^^
//│ res: error | int


:ex
(fun x -> succ x) false
//│ ╔══[ERROR] Type mismatch in application:
//│ ║  l.255: 	(fun x -> succ x) false
//│ ║         	^^^^^^^^^^^^^^^^^^^^^^^
//│ ╟── reference of type `false` is not an instance of type `int`
//│ ║  l.255: 	(fun x -> succ x) false
//│ ║         	                  ^^^^^
//│ ╟── Note: constraint arises from type reference:
//│ ║  l.3: 	def succ: int -> int
//│ ║       	          ^^^
//│ ╟── from reference:
//│ ║  l.255: 	(fun x -> succ x) false
//│ ║         	               ^
//│ ╟── ========= Additional explanations below =========
//│ ╟── [info] flowing from type `false`
//│ ╟── [info] flowing from reference of type `false`
//│ ║  l.255: 	(fun x -> succ x) false
//│ ║         	                  ^^^^^
//│ ╟── [info] flowing into variable of type `?a`
//│ ╟── [info] flowing into variable of type `int`
//│ ╟── [info] flowing into reference of type `int`
//│ ║  l.255: 	(fun x -> succ x) false
//│ ║         	               ^
//│ ╟── [info] flowing into type `int`
//│ ║  l.3: 	def succ: int -> int
//│ ║       	          ^^^
//│ ╟── [info] flowing into type `int`
//│ ║  l.3: 	def succ: int -> int
//│ ║       	          ^^^
//│ ╟── [info] flowing into type `int`
//│ ║  l.3: 	def succ: int -> int
//│ ╙──     	          ^^^
//│ res: error | int

// :d
:ex
rec def x = add x
//│ ╔══[ERROR] Type mismatch in binding of application:
//│ ║  l.291: 	rec def x = add x
//│ ║         	            ^^^^^
//│ ╟── application of type `int -> int` is not an instance of type `int`
//│ ╟── Note: constraint arises from reference:
//│ ║  l.291: 	rec def x = add x
//│ ║         	                ^
//│ ╟── ========= Additional explanations below =========
//│ ╟── [info] flowing from expression of type `int -> int`
//│ ╟── [info] flowing from application of type `int -> int`
//│ ║  l.291: 	rec def x = add x
//│ ║         	            ^^^^^
//│ ╟── [info] flowing from application of type `?a`
//│ ║  l.291: 	rec def x = add x
//│ ║         	            ^^^^^
//│ ╟── [info] flowing into expression of type `?x`
//│ ╟── [info] flowing into reference of type `int`
//│ ║  l.291: 	rec def x = add x
//│ ║         	                ^
//│ ╟── [info] flowing into type `int`
//│ ╙── [info] flowing into type `int`
//│ x: int -> int


def foo: int | string
//│ foo: int | string

:ex
succ foo
//│ ╔══[ERROR] Type mismatch in application:
//│ ║  l.320: 	succ foo
//│ ║         	^^^^^^^^
//│ ╟── type `string` is not an instance of `int`
//│ ║  l.316: 	def foo: int | string
//│ ║         	               ^^^^^^
//│ ╟── but it flows into reference with expected type `int`
//│ ║  l.320: 	succ foo
//│ ║         	     ^^^
//│ ╟── Note: constraint arises from type reference:
//│ ║  l.3: 	def succ: int -> int
//│ ║       	          ^^^
//│ ╟── ========= Additional explanations below =========
//│ ╟── [info] flowing from type `string`
//│ ║  l.316: 	def foo: int | string
//│ ║         	               ^^^^^^
//│ ╟── [info] flowing from type `int | string`
//│ ║  l.316: 	def foo: int | string
//│ ║         	         ^^^^^^^^^^^^
//│ ╟── [info] flowing from reference of type `int | string`
//│ ║  l.320: 	succ foo
//│ ║         	     ^^^
//│ ╟── [info] flowing into type `int`
//│ ║  l.3: 	def succ: int -> int
//│ ╙──     	          ^^^
//│ res: error | int



class A
class B
//│ Defined class A
//│ Defined class B

def ty00: ('a & A | 'b & B) -> ('a, 'b)
//│ ty00: (A & 'a | B & 'b) -> ('a, 'b,)

def ty11: ('a & A | 'a & B) -> ('a, 'a)
//│ ty11: ('a & (A | B)) -> ('a, 'a,)

ty11 = ty00
//│ (A & 'a | B & 'b) -> ('a, 'b,)
//│   <:  ty11:
//│ ('a & (A | B)) -> ('a, 'a,)

// :d
:ex
ty00 = ty11
//│ ('a & (A | B)) -> ('a, 'a,)
//│   <:  ty00:
//│ (A & 'a | B & 'b) -> ('a, 'b,)
//│ ╔══[ERROR] Type mismatch in def definition:
//│ ║  l.368: 	ty00 = ty11
//│ ║         	^^^^^^^^^^^
//│ ╟── type `B & 'b` does not match type `'a`
//│ ║  l.355: 	def ty00: ('a & A | 'b & B) -> ('a, 'b)
//│ ║         	                    ^^^^^^
//│ ╟── Note: constraint arises from type variable:
//│ ║  l.355: 	def ty00: ('a & A | 'b & B) -> ('a, 'b)
//│ ║         	                                ^^
//│ ╟── ========= Additional explanations below =========
//│ ╟── [info] flowing from expression of type `B & 'b`
//│ ╟── [info] flowing from type `B & 'b`
//│ ║  l.355: 	def ty00: ('a & A | 'b & B) -> ('a, 'b)
//│ ║         	                    ^^^^^^
//│ ╟── [info] flowing from type `B & 'b`
//│ ║  l.355: 	def ty00: ('a & A | 'b & B) -> ('a, 'b)
//│ ║         	          ^^^^^^^^^^^^^^^^^
//│ ╟── [info] flowing from type `B & 'b`
//│ ║  l.358: 	def ty11: ('a & A | 'a & B) -> ('a, 'a)
//│ ║         	          ^^^^^^^^^^^^^^^^^
//│ ╟── [info] flowing from expression of type `'a0`
//│ ╟── [info] flowing from type `'a0`
//│ ║  l.358: 	def ty11: ('a & A | 'a & B) -> ('a, 'a)
//│ ║         	                                ^^
//│ ╟── [info] flowing into type `'a`
//│ ║  l.355: 	def ty00: ('a & A | 'b & B) -> ('a, 'b)
//│ ║         	                                ^^
//│ ╙── [info] flowing into expression of type `'a`

